R/epidist_db.R

```diff
@@ -120,96 +127,436 @@ epidist_db <- function(disease,
+      "extdata",
+      "parameters.json",
+      package = "epiparameter",
+      mustWork = TRUE
+    )
+  )
+
+  # suppress individual <epidist> constructor messages
+  multi_epidist <- suppressMessages(lapply(params_json, .format_epidist))
+
+  # create and return <multi_epidist> class
+  structure(
+    multi_epidist,
+    class = "multi_epidist"
+  )
+}
+
+#' Filter a list of `<epidist>` objects by disease and epi distribution
+#'
+#' @param multi_epidist A list of `<epidist>` objects.
+#' @inheritParams epidist_db
+#'
+#' @return A list of `<epidist>` objects.
+#' @keywords internal
+#' @noRd
+.filter_epidist_db <- function(multi_epidist, disease, epi_dist) {
+  # copy of user input
+  disease_ <- disease
+  epi_dist_ <- epi_dist
+
+  # clean input strings
+  disease <- clean_disease(disease)
+  epi_dist <- clean_epi_dist(epi_dist)
+
+  # get valid options from db
+  disease_db <- vapply(
+    multi_epidist, function(x) x$disease$disease,
+    FUN.VALUE = character(1)
+  )
+  epi_dist_db <- vapply(
+    multi_epidist, function(x) x$epi_dist,
+    FUN.VALUE = character(1)
+  )
+  disease_db <- c("all", clean_disease(unique(disease_db)))
+  epi_dist_db <- c("all", clean_epi_dist(unique(epi_dist_db)))
+
+  # partial matching and custom error msg
+  tryCatch(
+    error = function(cnd) {
+      disease_str <- ifelse(
+        test = disease == "all", yes = "", no = paste(" for", disease_)
       )
-    } else {
-      lapply(edist, validate_epidist)
-
-      message(
-        "Returning ", nrow(eparam), " results that match the criteria ",
-        "(", sum(is_param), " are parameterised). \n",
-        "Use subset to filter by entry variables or ",
-        "single_epidist to return a single entry. \n",
-        "To retrieve the short citation for each use the ",
-        "'get_citation' function"
+      epi_dist_str <- ifelse(test = epi_dist == "all", yes = "", no = epi_dist_)
+      stop(
+        epi_dist_str, " distribution not available", disease_str,
+        call. = FALSE
+      )
+    },
+    {
+      disease <- match.arg(
+        arg = disease,
+        choices = disease_db,
+        several.ok = FALSE
+      )
+      epi_dist <- match.arg(
+        arg = epi_dist,
+        choices = epi_dist_db,
+        several.ok = FALSE
       )
     }
+  )
+
+  # filter based on disease
+  if (disease != "all") {
+    multi_epidist <- Filter(
+      f = function(x) clean_disease(x$disease$disease) == disease,
+      x = multi_epidist
+    )
   }
 
-  # return epidist
-  edist
+  # filter by epi dist
+  if (epi_dist != "all") {
+    multi_epidist <- Filter(
+      f = function(x) clean_epi_dist(x$epi_dist) == epi_dist,
+      x = multi_epidist
+    )
+  }
+
+  # return multi_epidist
+  multi_epidist
+}
+
+#' Format data from JSON database into `<epidist>` objects
+#'
+#' @param x A single entry (element/object) from the database.
+#'
+#' @return An `<epidist>` object.
+#' @keywords internal
+#' @noRd
+.format_epidist <- function(x) {
```

Comment:
As this functions is specifically formatting from the database and not a generalised method I think it is best to leave the name unchanged. This PR removes the `as_epidist()` function, but when it becomes clear that there are data structures that should be converted to `<epidist>` objects I will create an S3 generic `as_epidist()` and write methods for the relevant data structures.

Reply:
Currently there is no general method to create `<epidist>` objects from nested list objects. This would require writing data wrangling code which has to be inherently flexible to different input lists. This PR aims to reduce the amount of such code from the package until clear use-cases arise. (This assumes users don't just put an `epidist` class attribute on a list, which would technically be an `<epidist>` but would fail validation).